arch/arm64/src/imx9/imx9_usdhc.c

```diff
@@ -0,0 +1,3364 @@
+{
+  struct imx9_dev_s *priv = (struct imx9_dev_s *)dev;
+  DEBUGASSERT(priv != NULL && buffer != NULL && buflen > 0);
+  DEBUGASSERT(((uint64_t) buffer & 3) == 0);
+
+  /* Begin sampling register values */
+
+  imx9_sampleinit();
+  imx9_sample(priv, SAMPLENDX_BEFORE_SETUP);
+
+#if !defined(CONFIG_ARM64_DCACHE_DISABLE)
+  if (((uintptr_t)buffer & (ARMV8A_DCACHE_LINESIZE - 1)) != 0 ||
+       (buflen & (ARMV8A_DCACHE_LINESIZE - 1)) != 0)
+    {
+      /* The read buffer is not cache-line aligned. Read to an internal
+       * buffer instead.
+       */
+
+      up_invalidate_dcache((uintptr_t)priv->rxbuffer,
+                           (uintptr_t)priv->rxbuffer + priv->blocksize);
+
+      priv->unaligned_rx = true;
+    }
+  else
+    {
+      up_invalidate_dcache((uintptr_t)buffer,
+                           (uintptr_t)buffer + buflen);
+
+      priv->unaligned_rx = false;
+    }
+#endif
+
+  /* Save the destination buffer information for use by the interrupt
+   * handler
+   */
+
+  priv->buffer = (uint32_t *)buffer;
+  priv->remaining = buflen;
+
+  /* Then set up the SDIO data path */
+
+  imx9_dataconfig(priv, false, buflen, USDHC_DTOCV_MAXTIMEOUT);
+
+  /* Configure the RX DMA */
+
+  imx9_configxfrints(priv, USDHC_DMADONE_INTS);
+#if !defined(CONFIG_ARM64_DCACHE_DISABLE)
+  if (priv->unaligned_rx)
+    {
+      putreg32((uint64_t) priv->rxbuffer,
+               priv->addr + IMX9_USDHC_DSADDR_OFFSET);
+    }
+  else
+#endif
+    {
+      putreg32((uint64_t) priv->buffer,
+               priv->addr + IMX9_USDHC_DSADDR_OFFSET);
+    }
+
+  /* Sample the register state */
+
+  imx9_sample(priv, SAMPLENDX_AFTER_SETUP);
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: imx9_dmasendsetup
+ *
+ * Description:
+ *   Setup to perform a write DMA.  If the processor supports a data cache,
+ *   then this method will also make sure that the contents of the DMA memory
+ *   and the data cache are coherent.  For write transfers, this may mean
+ *   flushing the data cache.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   buffer - The memory to DMA into
+ *   buflen - The size of the DMA transfer in bytes
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_IMX9_USDHC_DMA
+static int imx9_dmasendsetup(struct sdio_dev_s *dev,
+                              const uint8_t *buffer, size_t buflen)
+{
+  struct imx9_dev_s *priv = (struct imx9_dev_s *)dev;
+  DEBUGASSERT(priv != NULL && buffer != NULL && buflen > 0);
+  DEBUGASSERT(((uint64_t) buffer & 3) == 0);
+
+  /* Begin sampling register values */
+
+  imx9_sampleinit();
+  imx9_sample(priv, SAMPLENDX_BEFORE_SETUP);
+
+  /* Save the source buffer information for use by the interrupt handler */
+
+#if !defined(CONFIG_ARM64_DCACHE_DISABLE)
+  priv->unaligned_rx = false;
```

Comment:
This looks odd, why set the unaligned_rx here in sendsetup?

Reply:
actually this is needed here.. clears the flag before a new transfer.

Reply:
Nope, this is dmasendsetup. The flag is correctly initialized in dmarecvsetup. This flag has nothing to do with dma sending.

Reply:
If I remove the flag whole system crashes. :)